### PR TITLE
Add popup and hover tooltips to warn if we can't jump to code

### PR DESCRIPTION
**Pull request by markerikson**
We're agreed that the right solution long-term is to precalculate "does this event have a valid jumpable location?" in a routine, but until then it's nice to have a tooltip warning you that you can't jump for this event.

I was able to cobble together both a 5-second popup and an on-hover version of the tooltip:

**[Loom demo here](https://www.loom.com/share/ffa8f8b315f74d648c4fccaf74d3ad82)**

To test this out:

- Open https://devtools-git-feature-fe-1221-jump-code-toast-recordreplay.vercel.app/)/recording/redux-todos-app-with-events--8835992a-2708-4c59-928e-85ea1f0216a5?point=32451855387679743845332202100359236&time=15602.54918032787
- Scroll down to the first two "Click" events, and click "Jump to Code"

image

### CODE REVIEW

File: packages/replay-next/src/hooks/useTooltip.tsx
Comment (bvaughn) [inline packages/replay-next/src/hooks/useTooltip.tsx:23]: This seems like an awkward mix of default and named exports. Why is this in the same file?

File: packages/replay-next/src/hooks/useTooltip.tsx
Comment (bvaughn) [inline packages/replay-next/src/hooks/useTooltip.tsx:36]: This feels awkward to me. I think maybe a more straight forward approach would be to update `useTooltip` to accept an optional flag `forceVisible` or something that would override the internal state when present.

```ts
export default function useTooltip({
  className,
  containerRef,
  delay = 0,
  forceVisible = false,
  position,
  tooltip,
}: {
  className?: string;
  containerRef?: RefObject<HTMLElement>;
  delay?: number;
  forceVisible?: boolean;
  position?: TooltipPosition;
  tooltip: ReactNode;
}): {
  onMouseEnter: (event: MouseEvent) => void;
  onMouseLeave: (event: MouseEvent) => void;
  onMouseMove: (event: MouseEvent) => void;
  tooltip: ReactNode;
} {
```

Then I don't think we'd need to fork at all?

**Review by bvaughn — CHANGES_REQUESTED** [review]

File: packages/replay-next/src/hooks/useTooltip.tsx
Comment (markerikson) [reply packages/replay-next/src/hooks/useTooltip.tsx:23]: This and the other comment are the same reason: I was going for the simplest thing that would work, you said it made sense to fork it, and given that we might re-combine them later it didn't seem worth putting this in another file.

File: packages/replay-next/src/hooks/useTooltip.tsx
Comment (markerikson) [reply packages/replay-next/src/hooks/useTooltip.tsx:36]: Okay, I'll give that a shot.

File: packages/replay-next/src/hooks/useTooltip.tsx
Comment (bvaughn) [reply packages/replay-next/src/hooks/useTooltip.tsx:23]: Not knowing what you had in mind, forking seemed reasonable. Seeing this proposed implementation though, it seems like overkill. I dunno 😄

File: packages/replay-next/src/hooks/useTooltip.tsx
Comment (markerikson) [reply packages/replay-next/src/hooks/useTooltip.tsx:36]: @bvaughn : ah, I remember why I really ended up copy-pasting this.

`useTooltip` _expects_ that you're using its `onMouseEnter/onMouseLeave` callbacks in order to get a reference to the target element.  It doesn't have a way to pass that in.

It was easier to fork the entire hook and swap that out to be an argument instead.

Comment (bvaughn) [reply packages/replay-next/src/hooks/useTooltip.tsx:36]: k. Fork it then, but I think it should be in its own file. Mixing named and default exports like this seems confusing to me.

Seems like we could share the _position_ logic between the two though.

Comment (bvaughn) [reply packages/replay-next/src/hooks/useTooltip.tsx:36]: I guess the _other_ options would be to change `useTooltip` to accept a `targetRef` prop but then that makes the regular "tooltip" use cases a little more verbose so maybe it's not worth it.

I dunno. End of a long week and my brain is feeling friend tbh.

Comment (markerikson) [reply packages/replay-next/src/hooks/useTooltip.tsx:36]: Okay. Personally I've shifted to not being a fan of default exports in the first place :)  But that's a separate discussion.

And yeah, extracting the positioning logic was what I started with, but that's also its own rabbit hole and I didn't want to deal with that for now.

Comment (bvaughn) [reply packages/replay-next/src/hooks/useTooltip.tsx:36]: > Personally I've shifted to not being a fan of default exports in the first place

Yeah. I think I've been coming around to that position myself, although default exports do have a nice benefit of speeding up file navigation for me in VS Code (e.g. Cmd+P -> type name of function).

**Comment by jasonLaster** [comment]
The UX seems fine
1. I would not change the cursor - we dont do that elsewhere and it's a bit awkard
2. It'd be nice to improve the placement of the tooltip. cc @jonbell-lot23 
3. If we're going with a tooltip, we probably dont need the 5 seconds, it could just show up on hover for a failed jump

**Comment by markerikson** [comment]
@jasonLaster : define "improve the placement of the tooltip"?

**Comment by bvaughn** [comment]
> If we're going with a tooltip, we probably dont need the 5 seconds, it could just show up on hover for a failed jump

If _that's_ the case then we could just use the original `useTooltip` hook. (Is that the case? I dunno what @jonbell-lot23 and @markerikson decided here.)

File: packages/replay-next/src/hooks/usePositionedTooltip.tsx
Comment (bvaughn) [inline packages/replay-next/src/hooks/usePositionedTooltip.tsx:38]: I think sharing state inside and outside of the hook this way is still a bit odd, but I can see why you do it with the hide-on-scroll behavior. 🤷🏼

**Review by bvaughn — APPROVED** [review]
Would be good to get clarification from @jonbell-lot23

**Comment by jasonLaster** [comment]
Would defer to Jon/Brian on alignment.

**Comment by jonbell-lot23** [comment]
Hey team! It's Saturday and I'm writing from a beach trip.

If it's cool with everyone, I'd like to hold off on this until next week. I can write more details later.

Quick summary:

1. Yeah, let's not change the cursor
2. The text of the button itself can contain the error, no tooltip needed

But there's no harm in holding off, so that'd be my preference to reduce back and forth.

**Comment by jonbell-lot23** [comment]
@markerikson Here's what I recommend ([Figma link](https://www.figma.com/file/mAotNZRv6M5X0zHIFKPzFY/Replay-Design-Doc?node-id=16420%3A208823&t=wRqsi8GkLSYOUsRL-4)):
image

1. No need for a tooltip, let's use the capsule itself to message
2. No custom cursor needed, we can use a standard cursor to show it's inert (see bottom left image)

I considered changing the forward/backward icon into something like an X for this state, but it doesn't work as well as keeping the forward/backward buttons. We're already juggling a lot of concepts here, so it's important to keep things consistent where we can.

**Review by jonbell-lot23 — APPROVED** [review]
Love this! Sorry for the runaround where we tried tooltips, but I think it's clear that this is great way to make it as clear yet simple as possible. Ship it!

**Pull request merged**
